kernel/syscall.c

```diff
@@ -1684,6 +1684,59 @@ long myst_syscall_readlink(const char* pathname, char* buf, size_t bufsiz)
     return ret;
 }
 
+long myst_syscall_readlinkat(
+    int dirfd,
+    const char* pathname,
+    char* buf,
+    size_t bufsiz)
+{
+    long ret = 0;
+    struct locals
+    {
+        char abspath[PATH_MAX];
+    };
+    struct locals* locals = NULL;
+
+    if (!pathname || !buf || !bufsiz)
+        ERAISE(-EINVAL);
+
+    /* If pathname is absolute, then ignore dirfd */
+    if (*pathname == '/' || dirfd == AT_FDCWD)
+    {
+        ret = myst_syscall_readlink(pathname, buf, bufsiz);
+    }
+    else if (*pathname == '\0')
+    {
+        /*
+         * ATTN: Since Linux 2.6.39, pathname can be an empty string, in which
```

Comment:
I think like fstatat we should implement the empty path case. The O_PATH issue documented in #489 can be addressed in a later PR.

Reply:
@vtikoo There is some nuance here.

`readlinkat` should deny empty `pathname` because `dirfd` should have never passed the condition check "dirfd should have been obtained using open with the O_PATH and O_NOFOLLOW flags" in our current implementation of open:
https://github.com/deislabs/mystikos/blob/d0ec79655a48ba54d0b69665d98a6c6304460ea7/ext2/ext2.c#L3357-L3358
In the above code, open a symlink with `O_NOFOLLOW` would fail with error `ELOOP`. It doesn't check if `O_PATH` is also specified. In `open` man page, it said:

**O_PATH** (since Linux 2.6.39)
> If pathname is a symbolic link and the O_NOFOLLOW flag is
              also specified, then the call returns a file descriptor
              referring to the symbolic link.  This file descriptor can
              be used as the dirfd argument in calls to fchownat(2),
              fstatat(2), linkat(2), and readlinkat(2) with an empty
              pathname to have the calls operate on the symbolic link.

Once #489 is fixed, I can submit another PR to enable empty path here.

Reply:
Ah thanks for clarifying, the code comment makes sense to me now.
Sounds good, I'll withdraw the suggestion.